blockprod/src/detail/block_maker.rs

```diff
@@ -0,0 +1,184 @@
+            .await
+            .log_err()?;
+        if let Ok(block) = block_check_result {
+            let block_id = block.get_id();
+            let block_submit_result = self
+                .chainstate_handle
+                .call_mut(|chainstate| {
+                    chainstate.process_block(block, chainstate::BlockSource::Local)
+                })
+                .await
+                .log_err()?;
+            if let Ok(_new_block_index) = block_submit_result {
+                log::info!(
+                "Success in submitting block {} at height {}. Exiting block maker at tip {} and height {}",
+                block_id,
+                self.current_tip_height.next_height(),
+                self.current_tip_id,
+                self.current_tip_height
+            );
+                return Ok(BlockSubmitResult::Success);
+            }
+        }
+
+        Ok(BlockSubmitResult::Failed)
+    }
+
+    /// Keeps trying to construct a new block, until one of two things happen:
+    /// 1. A new block is successfully created and is submitted to chainstate
+    /// 2. A new tip is now on chainstate, indicating that there's no point in continuing to mine/stake at that tip
+    pub async fn run(&mut self) -> Result<(), BlockProductionError> {
+        let accumulator = self.collect_transactions().await?;
+
+        loop {
+            let block = self.make_block(self.current_tip_id, &*accumulator)?;
+
+            match self.attempt_submit_new_block(block).await? {
+                BlockSubmitResult::Failed => (),
+                BlockSubmitResult::Success => break,
+            }
+
+            // attempt to receive new commands from the perpetual builder
+            let new_info = match self.block_maker_rx.try_recv() {
+                Ok(cmd) => cmd,
+                Err(e) => match e {
+                    // if there's nothing from the channel, then we can keep trying to build the block
+                    crossbeam_channel::TryRecvError::Empty => continue,
+                    // if the channel is lost, that means the perpetual builder is destroyed.
+                    // No point in continuing since it seems that the node exited.
+                    crossbeam_channel::TryRecvError::Disconnected => {
+                        log::error!("Block maker control channel lost. Exiting maker task on tip {} on best height {}", self.current_tip_id, self.current_tip_height);
+                        break;
+                    }
+                },
+            };
+
+            match new_info {
+                BlockMakerControlCommand::StopBecauseNewTip(block_id, _) => {
+                    // if there is a new tip, no point in continuing to mine this block
+                    if block_id != self.current_tip_id {
+                        break;
+                    }
+                }
+                BlockMakerControlCommand::JustStop => break,
+            }
+        }
```

Comment:
We will have to be careful not to starve other tasks with a busy loop like this. E.g. if `make_block` contains some intensive task like PoW mining or expensive block verification

Reply:
Unfortunately, this is how mining works. I already had discussions with Aaro about this and that we may have to introduce multiple executors to solve such problems. But... well, one problem at a time. I'll add a TODO to study this.